### PR TITLE
NET 1594 - Snapshot Agent Filename Should Include Consul Version / Datacenter

### PR DESCRIPTION
**Pull request by absolutelightning**
### Description

Adds flag -append-filename (which works on values version, dc, node and status) to consul snapshot save command.
Adding the flag -append-filename version,dc,node,status will add consul version, consul datacenter, node name and leader/follower
(status) in the file name given in the snapshot save command before the file extension.

### Testing & Reproduction steps
CI
Manual Testing -  
`make dev`
`./bin/consul agent -dev`
```
asheshvidyut@asheshvidyut-H2GX766V9T ~/test » ~/consul/bin/consul snapshot save -append-filename node,status,version,dc backup.tgz
Saved and verified snapshot to index 58
asheshvidyut@asheshvidyut-H2GX766V9T ~/test » ls
backup-1.17.0-dc1-asheshvidyut-H2GX766V9T-leader.tgz
```

### PR Checklist

* [x] updated test coverage
* [x] external facing docs updated
* [x] appropriate backport labels added
* [x] not a security concern

### CODE REVIEW

File: command/snapshot/save/snapshot_save.go [hunk 1]
```diff
@@ -58,6 +70,41 @@ func (c *cmd) Run(args []string) int {
 
 	// Create and test the HTTP client
 	client, err := c.http.APIClient()
+
+	appendFileNameFlags := strings.Split(c.appendFileNameFlag.String(), ",")
+
+	var agentSelfResponse map[string]map[string]interface{}
```
Comment (Ganeshrockz) [inline command/snapshot/save/snapshot_save.go:76]: This line can be removed

File: command/snapshot/save/snapshot_save.go [hunk 2]
```diff
@@ -23,17 +26,26 @@ func New(ui cli.Ui) *cmd {
...
+	UI                 cli.Ui
+	flags              *flag.FlagSet
+	http               *flags.HTTPFlags
+	help               string
+	appendFileNameFlag flags.StringValue
+}
+
+func (c *cmd) getAppendFileNameFlag() *flag.FlagSet {
+	fs := flag.NewFlagSet("", flag.ContinueOnError)
+	fs.Var(&c.appendFileNameFlag, "append-filename", "Append filename flag takes two possible values. "+
```
Comment (Ganeshrockz) [inline command/snapshot/save/snapshot_save.go:38]: I guess `append-to-filename` sounds better semantically. WDYT?

File: command/snapshot/save/snapshot_save.go [hunk 3]
```diff
@@ -23,17 +26,26 @@ func New(ui cli.Ui) *cmd {
...
+		"1. version, 2. dc. It appends consul version and datacenter to filename given in command")
```
Comment (Ganeshrockz) [inline command/snapshot/save/snapshot_save.go:39]: Should we add validation to check if other fields or present?

**Review by Ganeshrockz — APPROVED** [review]
Looks good to me. Added some minor comments but those shouldn't be blockers.

File: command/snapshot/save/snapshot_save.go (hunk 2 above)
Comment (absolutelightning) [reply command/snapshot/save/snapshot_save.go:38]: This was given by @david-yu . So kept it that way. I am fine with both.

Comment (absolutelightning) [reply command/snapshot/save/snapshot_save.go:38]: https://hashicorp.slack.com/archives/C055CK3BMR8/p1693461179291959?thread_ts=1693397494.647169&cid=C055CK3BMR8

File: command/snapshot/save/snapshot_save.go (hunk 3 above)
Comment (absolutelightning) [reply command/snapshot/save/snapshot_save.go:39]: I prefer giving results based on given flags.

**Comment by david-yu** [comment]
Thanks could you also update the usage docs and provide an example in the doc of how to use the flag? It would be good to provide the supported options available such as dc, version. 

Docs: https://developer.hashicorp.com/consul/commands/snapshot

**Comment by david-yu** [comment]
It would also perhaps be good to indicate whether the server is a leader or follower and node name, i.e. 

```
consul snapshot save -append-filename version,dc,node,status backup.snap 
```

Which should follow the columns described here: https://developer.hashicorp.com/consul/commands/operator/raft#list-peers

**Comment by absolutelightning** [comment]
> It would also perhaps be good to indicate whether the server is a leader or follower and node name, i.e.
> 
> ```
> consul snapshot save -append-filename version,dc,node,status backup.snap 
> ```
> 
> Which should follow the columns described here: https://developer.hashicorp.com/consul/commands/operator/raft#list-peers

sure

**Comment by absolutelightning** [comment]
@david-yu - updated docs and added node, status in flag as well.

**Review by david-yu — APPROVED** [review]

File: website/content/commands/snapshot/save.mdx [hunk 4]
```diff
@@ -49,6 +49,10 @@ Usage: `consul snapshot save [options] FILE`
 
 @include 'http_api_options_server.mdx'
 
+- `-append-filename=<value>` - Value can be - version,dc,node,status
+Adds consul version, datacenter name, node name, and status (leader/follower)
+to the file name before the extension separated by `-`
+
 ## Examples
```
Comment (david-yu) [inline website/content/commands/snapshot/save.mdx:56]: Could you provide an example using the append-filename flag here?

Comment (absolutelightning) [reply website/content/commands/snapshot/save.mdx:56]: done

File: command/snapshot/save/snapshot_save.go (hunk 2 above)
Comment (JadhavPoonam) [inline command/snapshot/save/snapshot_save.go:38]: This description is a bit confusing to me. Can we rephrase this to something like?
```
Append filename flag supports the following comma-separated arguments. 1. version, 2. dc. 3. node 4. status. It appends these values to the filename provided in the command
```

**Review by JadhavPoonam — APPROVED** [review]
Left a comment about the help text otherwise looks fine.

**Comment by absolutelightning** [comment]
@david-yu Are backports required in this one?

**Comment by david-yu** [comment]
Yes, I added 1.16.x, 1.15.x, and 1.14.x
